execution/tvf/percentile_watermark_generator.go

```diff
@@ -0,0 +1,300 @@
+	}
+
+	return currentWatermark.AsTime(), nil
+}
+
+func (s *PercentileWatermarkGeneratorStream) Next(ctx context.Context) (*execution.Record, error) {
+	srcRecord, err := s.source.Next(ctx)
+	if err != nil {
+		if err == execution.ErrEndOfStream {
+			return nil, execution.ErrEndOfStream
+		}
+		return nil, errors.Wrap(err, "couldn't get source record")
+	}
+
+	timeValue := srcRecord.Value(s.timeField)
+	if timeValue.GetType() != octosql.TypeTime {
+		return nil, fmt.Errorf("couldn't get time field '%v' as time, got: %v", s.timeField.String(), srcRecord.Value(s.timeField))
+	}
+
+	tx := storage.GetStateTransactionFromContext(ctx)
+	watermarkStorage := storage.NewValueState(tx.WithPrefix(percentileWatermarkPrefix))
+	eventsStorage := storage.NewDeque(tx.WithPrefix(percentileWatermarkEventsPrefix))
+	eventsCountStorage := storage.NewMap(tx.WithPrefix(percentileWatermarkEventsCountPrefix))
+	eventsSeenStorage := storage.NewValueState(tx.WithPrefix(percentileWatermarkEventsSeenPrefix))
+
+	// Extracting eventsSeen value (remember: this equals to number of events seen before watermark update)
+	var eventsSeen octosql.Value
+	err = eventsSeenStorage.Get(&eventsSeen)
+	if err == storage.ErrNotFound {
+		eventsSeen = octosql.MakeInt(0)
+	} else if err != nil {
+		return nil, errors.Wrap(err, "couldn't get events seen count")
+	}
+	eventsSeen = octosql.MakeInt(eventsSeen.AsInt() + 1) // Adding +1 as we've just extracted next record from source
+
+	// Adding newest event
+	err = addNewestEvent(timeValue, tx)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't add newest event to storage")
+	}
+
+	eventsLength, err := eventsStorage.Length()
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get events deque length")
+	}
+
+	// In this situation we need to remove oldest event => this doesn't occur only when the oldest event seen is the first ever and length = <events> (update watermark but DON'T remove oldest one)
+	if eventsLength > s.events {
+		err := removeOldestEvent(tx)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't remove oldest event from storage")
+		}
+	}
+
+	// Updating watermark; enough events stored in deque AND frequency level reached
+	// '>=' for the case when we doesn't remove oldest event (explained above)
+	if eventsLength >= s.events && eventsSeen.AsInt() >= s.frequency {
+
+		// Resetting events seen
+		eventsSeen = octosql.MakeInt(0)
+
+		// Below declaration equals to (percentile*1000 / 100*1000) = (events - wP) / events
+		// Using (events - wP) instead of wP, because percentile of 80% means, that 80% events can be BIGGER than watermark value
+		// so watermark position is at 20% of all events
+		// Multiplying percentile by 1000 to allow float values like 99.9%
+		watermarkPosition := float64((100000-int(s.percentile*1000))*s.events) / 100000 // represents position (from the left) of event in sorted list that will become new watermark
```

Comment:
Why the float64 cast? The position always has to be a whole number, so an integer will suffice.